lib/trino-orc/src/main/java/io/trino/orc/metadata/statistics/StringStatisticsBuilder.java

```diff
@@ -158,16 +176,82 @@ public ColumnStatistics buildColumnStatistics()
         return stringStatisticsBuilder.buildStringStatistics();
     }
 
-    private Slice dropStringMinMaxIfNecessary(Slice minOrMax)
+    private Slice computeStringMinMax(Slice minOrMax, boolean isMin)
     {
-        if (minOrMax == null || minOrMax.length() > stringStatisticsLimitInBytes) {
+        if (minOrMax == null || (!shouldCompactMinMax && minOrMax.length() > stringStatisticsLimitInBytes)) {
             return null;
         }
+        if (minOrMax.length() > stringStatisticsLimitInBytes) {
+            if (isMin) {
+                return StringCompactor.truncateMin(minOrMax, stringStatisticsLimitInBytes);
+            }
+            else {
+                return StringCompactor.truncateMax(minOrMax, stringStatisticsLimitInBytes);
+            }
+        }
 
         // Do not hold the entire slice where the actual stats could be small
         if (minOrMax.isCompact()) {
             return minOrMax;
         }
         return Slices.copyOf(minOrMax);
     }
+
+    static final class StringCompactor
+    {
+        private StringCompactor() {}
+
+        public static Slice truncateMin(Slice slice, int maxBytes)
+        {
+            checkArgument(slice.length() > maxBytes);
+            if (maxBytes == 0) {
+                return null;
+            }
+            int lastIndex = findLastCharacterInRange(slice, maxBytes);
+            return slice.slice(0, lastIndex);
+        }
+
+        public static Slice truncateMax(Slice slice, int maxBytes)
+        {
+            if (maxBytes == 0) {
+                return null;
+            }
+            int firstRemovedCharacterIndex = findLastCharacterInRange(slice, maxBytes);
```

Comment:
this will throw (`IllegalArgumentException("Provided byte array is not a valid utf8 string")`) when input is eg 4-byte utf8 sequence, and maxBytes=3 (please add a test)

let's

- change so that `findLastCharacterInRange` returns `-1` (or perhaps Optional.empty) in such case 
  - return early when we get -1 here
- remove `if (maxBytes == 0)` above, as it becomes obsolete

Reply:
propose something else?